File: daily_problems/2026/03/0312/personal_submission/cf105453e_yawn_sean.cpp
#include <bits/stdc++.h>
// #pragma GCC optimize("O3,Ofast,unroll-loops")
// #include "atcoder/all"

using namespace std;
auto rng = mt19937(random_device()());
auto rngl = mt19937_64(random_device()());

// Let's unordered_set a bit and flow!
// I came, I divided, I conquered!
// using namespace __gnu_cxx;
// using namespace __gnu_pbds;

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);

	int n;
	cin >> n;

	vector<vector<int>> grid(n, vector<int>(n));
	for (auto &x: grid) for (auto &y: x) cin >> y;

	vector<int> cur(n), vis(n, 0);
	for (int i = 0; i < n; i ++) cur[i] = grid[i][i];

	int ans = 0;

	for (int i = 0; i < n; i ++) {
		int chosen = -1;
		for (int j = 0; j < n; j ++) {
			if (!vis[j] && (chosen == -1 || cur[j] < cur[chosen])) {
				chosen = j;
			}
		}

		ans += cur[chosen];
		vis[chosen] = 1;

		for (int j = 0; j < n; j ++) {
			cur[j] = min(cur[j], grid[chosen][j]);
		}
	}

	cout << ans;

	return 0;
}

File: daily_problems/2026/03/0312/personal_submission/cf105453e_yawn_sean.py
# Submission link: https://codeforces.com/gym/105453/submission/366243186
def main(): 
    n = II()
    grid = [LII() for _ in range(n)]
    
    ans = 0
    
    vis = [0] * n
    cur = [grid[i][i] for i in range(n)]
    
    for _ in range(n):
        chosen = -1
        
        for i in range(n):
            if not vis[i] and (chosen == -1 or cur[i] < cur[chosen]):
                chosen = i
        
        ans += cur[chosen]
        vis[chosen] = 1
        
        for j in range(n):
            cur[j] = fmin(cur[j], grid[chosen][j])
    
    print(ans)